src/examples/mnist/MNIST_SP.cpp

```diff
@@ -0,0 +1,240 @@
+        std::reverse((char*) &magic_number,      (char*) &magic_number + 4);
+        std::reverse((char*) &number_of_labels,  (char*) &number_of_labels + 4);
+    }
+    if(magic_number != 0x00000801) {
+        cerr << "ERROR: MNIST data is compressed or corrupt" << endl;
+        exit(1);
+    }
+    vector<UInt> retval;
+    for(int i = 0; i < number_of_labels; ++i) {
+        unsigned char label = 0;
+        file.read( (char*) &label, 1);
+        retval.push_back((UInt) label);
+    }
+    return retval;
+}
+
+
+vector<UInt*> read_mnist_images(string path) {
+    ifstream file(path);
+    if( !file.is_open() ) {
+        cerr << "ERROR: Failed to open file " << path << endl;
+        exit(1);
+    }
+    int magic_number     = 0;
+    int number_of_images = 0;
+    int n_rows           = 0;
+    int n_cols           = 0;
+    file.read( (char*) &magic_number,     4);
+    file.read( (char*) &number_of_images, 4);
+    file.read( (char*) &n_rows,           4);
+    file.read( (char*) &n_cols,           4);
+    if(magic_number != 0x00000803) {
+        std::reverse((char*) &magic_number,      (char*) &magic_number + 4);
+        std::reverse((char*) &number_of_images,  (char*) &number_of_images + 4);
+        std::reverse((char*) &n_rows,            (char*) &n_rows + 4);
+        std::reverse((char*) &n_cols,            (char*) &n_cols + 4);
+    }
+    if(magic_number != 0x00000803) {
+        cerr << "ERROR: MNIST data is compressed or corrupt" << endl;
+        exit(1);
+    }
+    NTA_ASSERT(n_rows == 28);
+    NTA_ASSERT(n_cols == 28);
+    UInt img_size = n_rows * n_cols;
+    vector<UInt*> retval;
+    for(int i = 0; i < number_of_images; ++i) {
+        auto data_raw = new unsigned char[img_size];
+        file.read( (char*) data_raw, img_size);
+        // Copy the data into an array of UInt's
+        auto data = new UInt[img_size];
+        // auto data = new UInt[2 * img_size];
+        // Apply a threshold to the image, yielding a B & W image.
+        for(UInt pixel = 0; pixel < img_size; pixel++) {
+            data[pixel] = data_raw[pixel] >= 128 ? 1 : 0;
+            // data[2 * pixel] = data_raw[pixel] >= 128 ? 1 : 0;
+            // data[2 * pixel + 1] = 1 - data[2 * pixel];
+        }
+        retval.push_back(data);
+        delete[] data_raw;
+    }
+    return retval;
+}
+
+
+int main(int argc, char **argv) {
+  UInt verbosity = 1;
+  auto train_dataset_iterations = 1u;
+  int opt;
+  while ( (opt = getopt(argc, argv, "tv")) != -1 ) {  // for each option...
+    switch ( opt ) {
+      case 't':
+          train_dataset_iterations += 1;
+        break;
+      case 'v':
+          verbosity = 1;
+        break;
+      case '?':
+          cerr << "Unknown option: '" << char(optopt) << "'!" << endl;
+        break;
+    }
+  }
+
+  SDR input({28, 28, 1});
+  SpatialPoolerExtended sp(
+    /* inputDimensions */           input.dimensions,
+    /* columnDimensions */          {10, 10, 120},
```

Comment:
please document the impact of the 3rd dimension on performance (as discussed in the PR thread: performance-vs-time,)